### ISSUE TITLE
Week 2: Vision Transformers

### ISSUE THREAD

**Issue by andreifurdui** [andreifurdui/neural-stack#6]
## Week 2 Practical: From Pixels to Tokens

**Goal**: Build a complete Vision Transformer that converts images into token sequences, applies transformer blocks, and outputs class predictions. Understand computational tradeoffs through rigorous analysis.

**Target Time**: 11 Hours  
**Location**: `src/vision_transformer.py` and `notebooks/02_vit_architecture.ipynb`

---

## Learning Objectives
- [ ] Convert 2D images into 1D sequences via patch embeddings
- [ ] Implement learnable positional encodings for spatial awareness
- [ ] Build complete minimal ViT architecture end-to-end
- [ ] Quantify computational complexity as function of patch size
- [ ] Empirically validate design choices through ablation studies

---

## Phase 1: Patch Embedding Layer (3 hours)

**Implement**: `PatchEmbedding(nn.Module)` that converts images to token sequences.

### Requirements

**Constructor parameters:**
- `image_size` (int): Input image dimension (assume square)
- `patch_size` (int): Patch dimension (must divide `image_size` evenly)
- `in_channels` (int): Number of input channels (3 for RGB)
- `embed_dim` (int): Output embedding dimension

**Forward pass:**
```python
# Input:  (B, C, H, W) - Batch of images
# Output: (B, N, D) - Batch of token sequences
#   where N = (H/P) * (W/P) = number of patches
#         D = embed_dim
```

### Task Checklist

- [x] **Patch Extraction:** Use `nn.Conv2d` with `kernel_size=patch_size` and `stride=patch_size` to extract non-overlapping patches in one operation
- [x] **Flatten Spatial Dims:** Reshape from `(B, D, H/P, W/P)` to `(B, D, N)` then transpose to `(B, N, D)`
- [x] **Learnable Positional Embeddings:** Create `nn.Parameter` of shape `(1, N, D)` initialized with `torch.randn` scaled by `0.02`
- [x] **Class Token:** Create learnable `nn.Parameter` of shape `(1, 1, D)`, prepend to sequence → final shape `(B, N+1, D)`
- [x] Add positional embeddings to all tokens (including CLS)

### Success Criteria

✓ **Correctness:** For `image_size=32, patch_size=4`, produces `(B, 65, D)` output (64 patches + 1 CLS token)  
✓ **Efficiency:** No explicit loops over patches—single conv operation extracts all patches  
✓ **Documentation:** Docstring explains why conv with stride=patch_size works as patch extractor

---

## Phase 2: Complete ViT Architecture (4 hours)

**Implement**: `VisionTransformer(nn.Module)` that composes all components.

### Requirements

**Constructor parameters:**
- `image_size`, `patch_size`, `in_channels`, `embed_dim` (for patch embedding)
- `num_heads`, `num_layers`, `mlp_ratio`, `dropout` (for transformer)
- `num_classes` (for classification head)

**Architecture:**
```
Input Image → PatchEmbedding → [TransformerBlock] × L → CLS Token → MLP Head → Logits
```

### Task Checklist

- [ ] **Reuse Week 1:** Import your `MultiHeadAttention` from Week 1 practical
- [ ] **Transformer Block:** Implement `TransformerBlock(nn.Module)` with:
  - Layer Norm → Multi-Head Attention → Residual Connection
  - Layer Norm → MLP (Linear → GELU → Dropout → Linear) → Residual Connection
  - MLP hidden dim = `embed_dim * mlp_ratio` (typically 4)
- [ ] **Stack Blocks:** Use `nn.ModuleList` to create `num_layers` transformer blocks
- [ ] **Classification Head:** Extract CLS token `[:, 0, :]`, apply Layer Norm, then `nn.Linear(embed_dim, num_classes)`
- [ ] **Forward Pass:** Chain all components, return logits of shape `(B, num_classes)`

### Success Criteria

✓ **Architecture:** Matches ViT paper (Layer Norm before attention/MLP, residual connections)  
✓ **Dimensionality:** Processes CIFAR-10 `(3, 32, 32)` → outputs `(B, 10)` logits  
✓ **Integration:** Successfully uses Week 1's attention without modification

---

## Phase 3: Computational Complexity Analysis (2 hours)

**Goal**: Understand the quadratic relationship between patch size and computation.

### Task Checklist

- [ ] **Calculate for 3 configurations** (CIFAR-10: 32×32×3, 6 layers, 8 heads, embed_dim=512):

| Patch Size | Num Patches | Tokens (w/ CLS) | Attention FLOPs/Layer | Total Params |
|------------|-------------|-----------------|----------------------|--------------|
| 4×4        | 64          | 65              | ?                    | ?            |
| 8×8        | 16          | 17              | ?                    | ?            |
| 16×16      | 4           | 5               | ?                    | ?            |

- [ ] **Attention complexity:** For N tokens, attention is O(N²·D). Calculate actual FLOPs: `4 * N² * D * num_heads` per layer
- [ ] **Parameter count:** Use `sum(p.numel() for p in model.parameters())`
- [ ] **Memory estimate:** Activations ≈ `batch_size * num_layers * N * embed_dim * 4 bytes`

### Success Criteria

✓ **Quantitative:** Completed table showing 4×4 patches have ~16× more attention ops than 16×16  
✓ **Insight:** Written explanation: "Why does the ViT paper use 16×16 patches for ImageNet but smaller patches might work for CIFAR-10?"  
✓ **Tradeoff:** Can articulate: smaller patches = more spatial detail but quadratic cost increase

---

## Phase 4: Ablation Study (2 hours)

**Goal**: Empirically validate that each component is necessary.

### Experiments to Run

**Setup:** Use CIFAR-10, patch_size=8, 4 layers, train for 10 epochs each (~5 min per run).

1. **Baseline:** Full ViT with positional embeddings
2. **No Positional Encoding:** Remove `self.pos_embed` addition
3. **No CLS Token:** Use mean pooling over all tokens instead
4. **Random Initialization Only:** Don't train, just test random model

### Task Checklist

- [ ] Implement simple training loop (CrossEntropyLoss, Adam optimizer)
- [ ] Track validation accuracy for each configuration
- [ ] **Visualize:** Plot 4 learning curves on same axes
- [ ] **Analyze:** Record final validation accuracy

### Expected Results

- **Baseline:** Should reach ~45-55% accuracy (untrained transformers need more data)
- **No Pos Encoding:** ~10-20% drop (model can't distinguish spatial positions)
- **No CLS:** ~5-10% drop (mean pooling loses expressiveness)
- **Random:** ~10% accuracy (chance level for 10 classes)

### Success Criteria

✓ **Empirical:** Positional embeddings provide clear accuracy improvement  
✓ **Visualization:** Learning curves clearly show baseline > ablations > random  
✓ **Understanding:** Can explain why each component helps (position info, learnable aggregation)

---

## Final Deliverable

**Submit:** Jupyter notebook containing:

1. **Working code:** `PatchEmbedding` and `VisionTransformer` classes with docstrings
2. **Complexity table:** Completed with all calculations shown
3. **Ablation results:** Learning curves plot + accuracy table
4. **Written analysis:** (≤500 words)
   - Why does removing positional embeddings hurt performance?
   - When would you choose 4×4 vs 16×16 patches?
   - How does ViT's architecture differ from CNNs in handling spatial information?

---

## Self-Assessment Questions

Answer without looking up:

1. **Calculation:** For 224×224 image, 16×16 patches, how many tokens? What's attention complexity?
2. **Architecture:** What happens if you swap Layer Norm position (post-attention vs pre-attention)?
3. **Design:** Why use learnable positional embeddings instead of fixed sinusoidal (like original Transformer)?
4. **Scaling:** If you 2× the image resolution, how much does computation increase?

---

## Common Pitfalls

⚠️ **Dimension mismatch:** Ensure `image_size % patch_size == 0`  
⚠️ **Positional embedding shape:** Must be `(1, N+1, D)` to include CLS token  
⚠️ **CLS token indexing:** Extract `output[:, 0, :]` not `output[:, -1, :]`  
⚠️ **Attention mask:** Not needed for ViT (full self-attention, no padding)

---

## Advanced Extensions (Optional)

If you finish early:
- Implement 2D positional embeddings (separate for row/col) and compare to 1D
- Visualize attention maps from different layers (early vs late layers)
- Add learnable temperature scaling to attention softmax
